### DISCUSSION TITLE
Additional detail on package use of nsplit?

### DISCUSSION THREAD

**Discussion by glyden** [kogalur/randomForestSRC#374] [Q&A] [answered]
1. Regarding split-points for categorical predictors, the `rfsrc()` Details note: "there is an optimization check to ensure number of splits attempted is not greater than number of cases in a node or the value of `nsplit`". Does this mean that `nsplit` possible split-points are randomly selected from all possible splits of the categorical predictor's levels?

2. For continuous predictors, are the `nsplit` split-points randomly selected from the empirical distribution of the predictor or the set of unique values?

3. Should `nsplit` be tuned similar to `mtry` and `nodesize` or should analysts simply use the largest value for `nsplit` that is computationally feasible?

Thanks!

**Comment by ishwaran** [comment, marked as answer]
1. Splits on categorical variables are made in terms of complementary pairs.  Consider a factor having three levels {A,B,C}.  Then complementary pairs are {A} versus {B,C} or {A,B} versus {C} or {A,C} versus {B}.  The number of such pairs grows exponentially in terms of the number of levels of a factor, therefore we cap this at the `nsplit` value (or the number of cases in a node) in the setting where we might have a factor with many levels.
2. It's selected from the unique values.
3. I don't believe there is a much advantage in tuning `nsplit`.  It can make a difference in *some* cases but primarily the biggest difference is computational speed.

**Reply by glyden** [reply]
Thank you for such a helpful response! I have two quick follow-up questions.

1. Is there also a check to ensure that the `nsplit` selected split-points would each result in a "meaningful" split, i.e., observations on both sides of the split-point, for the particular node?

2. Does the empirical distribution of a continuous predictor play a role? Let me give an example. I have a continuous predictor that is 75% 0. The other 25% of values range from >0 to 100. One can imagine that >0 would be a very meaningful split-point for this predictor. But if `nsplit=5` possible split-points are randomly selected from the _unique_ set of values, it is very likely these possible split-points will not be 0. So, does the empirical distribution play a role in the random selection of split-points, or are the split-points randomly selected from the unique set of values?

**Comment by ishwaran** [comment]
1. Yes.  The code always produces binary splits (null daughters are not allowed; in fact we also have a purity check in place that stops splitting when purity of the response is obtained).
2. The empirical distribution does _not_  play a role.  Splitting randomly on the unique observed values in a node rather than splitting on the empirical distribution avoids this from happening.  We split on the observed unique values.  For example if the observed predictor values in a node to be split is comprised of  75% values 0, 25 % values 1, we select {0,1} with equal probability.

**Reply by glyden** [reply]
That answers my question perfectly, thank you! This is great to know.
